apis/actions.github.com/v1alpha1/autoscalingrunnerset_types.go

```diff
@@ -80,6 +85,124 @@ type ProxyConfig struct {
 
 	// +optional
 	HTTPS *ProxyServerConfig `json:"https,omitempty"`
+
+	// +optional
+	NoProxy []string `json:"noProxy,omitempty"`
+}
+
+func (c *ProxyConfig) ProxyFunc(secretFetcher func(string) (*corev1.Secret, error)) (func(*http.Request) (*url.URL, error), error) {
+	config := &httpproxy.Config{
+		NoProxy: strings.Join(c.NoProxy, ","),
+	}
+
+	if c.HTTP != nil {
+		u, err := url.Parse(c.HTTP.Url)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.HTTP.CredentialSecretRef != "" {
+			secret, err := secretFetcher(c.HTTP.CredentialSecretRef)
+			if err != nil {
+				return nil, err
+			}
+
+			u.User = url.UserPassword(
+				string(secret.Data["username"]),
+				string(secret.Data["password"]),
+			)
+		}
+
+		config.HTTPProxy = u.String()
+	}
+
+	if c.HTTPS != nil {
+		u, err := url.Parse(c.HTTPS.Url)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.HTTPS.CredentialSecretRef != "" {
+			secret, err := secretFetcher(c.HTTPS.CredentialSecretRef)
+			if err != nil {
+				return nil, err
+			}
+
+			u.User = url.UserPassword(
+				string(secret.Data["username"]),
+				string(secret.Data["password"]),
+			)
+		}
+
+		config.HTTPSProxy = u.String()
+	}
+
+	proxyFunc := func(req *http.Request) (*url.URL, error) {
+		return config.ProxyFunc()(req.URL)
+	}
+
+	return proxyFunc, nil
+}
+
+func (c *ProxyConfig) EnvVars(secretFetcher func(string) (*corev1.Secret, error)) ([]corev1.EnvVar, error) {
+	var vars []corev1.EnvVar
+
+	if c.HTTP != nil {
+		u, err := url.Parse(c.HTTP.Url)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.HTTP.CredentialSecretRef != "" {
+			secret, err := secretFetcher(c.HTTP.CredentialSecretRef)
+			if err != nil {
+				return nil, err
+			}
+
+			u.User = url.UserPassword(
+				string(secret.Data["username"]),
+				string(secret.Data["password"]),
+			)
+		}
+
+		vars = append(vars, corev1.EnvVar{
+			Name:  "http_proxy",
+			Value: u.String(),
+		})
+	}
+
+	if c.HTTPS != nil {
+		u, err := url.Parse(c.HTTPS.Url)
+		if err != nil {
+			return nil, err
+		}
+
+		if c.HTTPS.CredentialSecretRef != "" {
+			secret, err := secretFetcher(c.HTTPS.CredentialSecretRef)
+			if err != nil {
+				return nil, err
+			}
+
+			u.User = url.UserPassword(
+				string(secret.Data["username"]),
+				string(secret.Data["password"]),
+			)
+		}
+
+		vars = append(vars, corev1.EnvVar{
+			Name:  "https_proxy",
+			Value: u.String(),
+		})
+	}
+
+	if len(c.NoProxy) > 0 {
+		vars = append(vars, corev1.EnvVar{
+			Name:  "no_proxy",
+			Value: strings.Join(c.NoProxy, ","),
+		})
+	}
+
+	return vars, nil
 }
 
 type ProxyServerConfig struct {
```

Comment:
I know this is kind of too late for this type of suggestion, but stilling want to bring it up...
What if we change `ProxyServerConfig` to be
```go
type ProxyServerConfig struct {
  // +optional
  AnonymousUrl string `json:"anonymousUrl,omitempty"`

  // +optional
  AuthenticatedUrlSecretRef string `json:"AuthenticatedUrlSecretRef,omitempty"`
}
```

And we ask the customer to provide the wellformed auth URL as the data in the secret.

I was trying to think if the customer already has an auth proxy, what kind of secret they may already have, can we reuse that?

Sorry about the extra thoughts.

Reply:
I actually like this, and if there's a time to make this kind of changes is now 👍 I genuinely don't know what would be more comfortable/less confusing from a user perspective. Just putting both cases here to see them next to each other

```go
type ProxyServerConfig struct {
  // Required
  Url string `json:"url,omitempty"`

  // +optional
  CredentialSecretRef string `json:"credentialSecretRef,omitempty"`
}

// vs

type ProxyServerConfig struct {
  // +optional
  AnonymousUrl string `json:"anonymousUrl,omitempty"`

  // +optional
  AuthenticatedUrlSecretRef string `json:"AuthenticatedUrlSecretRef,omitempty"`
}
```

I like that in the second case we don't deal with building the URL ourselves, although the first one seems less confusing to me 🤔

Reply:
Isn't the 2nd definition contained in the first one? I mean, can't one just use `Url` as either `AnonymousUrl` or `AuthenticatedUrlSecretRef` and leave `CredentialSecretRef` empty?

Reply:
We decided to go with the current spec.